htlcswitch/link_test.go

```diff
@@ -3913,3 +3913,377 @@ func restartLink(aliceChannel *lnwallet.LightningChannel, aliceSwitch *Switch,
+		t.Fatalf("Expected 1 htlc sig, got %d",
+			len(commitSig.HtlcSigs))
+	}
+
+	// Alice receives signature covering htlc1.
+	aliceLink.HandleChannelUpdate(commitSig)
+
+	// Generate a second payment.
+	invoice2, htlc2, err := generatePayment(htlcAmt, htlcAmt, 144, blob)
+	if err != nil {
+		t.Fatalf("unable to create payment: %v", err)
+	}
+
+	// Alice must also be expecting this one.
+	err = coreLink.cfg.Registry.(*mockInvoiceRegistry).AddInvoice(*invoice2)
+	if err != nil {
+		t.Fatalf("unable to add invoice to registry: %v", err)
+	}
+
+	htlc2.ID = 1
+	_, err = bobChannel.AddHTLC(htlc2, nil)
+	if err != nil {
+		t.Fatalf("bob failed adding htlc: %v", err)
+	}
+
+	// Alice receives htlc2 from Bob.
+	aliceLink.HandleChannelUpdate(htlc2)
+
+	// Since Bob has sent a new commitment signature, Alice should first
+	// respond with a revocation.
+	var msg lnwire.Message
+	select {
+	case msg = <-aliceMsgs:
+	case <-time.After(15 * time.Second):
+		t.Fatalf("did not receive message")
+	}
+
+	rev, ok := msg.(*lnwire.RevokeAndAck)
+	if !ok {
+		t.Fatalf("expected RevokeAndAck, got %T", msg)
+	}
+
+	// This revocation will also make Alice ack htlc1.
+	_, _, _, err = bobChannel.ReceiveRevocation(rev)
+	if err != nil {
+		t.Fatalf("bob failed receiving revocation: %v", err)
+	}
+
+	// Alice should also send a commitment signature for the new state,
+	// covering htlc1.
+	select {
+	case msg = <-aliceMsgs:
+	case <-time.After(15 * time.Second):
+		t.Fatalf("did not receive message")
+	}
+
+	comSig, ok := msg.(*lnwire.CommitSig)
+	if !ok {
+		t.Fatalf("expected CommitSig, got %T", msg)
+	}
+
+	// Make sure this signature covers htlc1.
+	if len(comSig.HtlcSigs) != 1 {
+		t.Fatalf("expected 1 htlc sigs, got %d",
+			len(comSig.HtlcSigs))
+	}
+	err = bobChannel.ReceiveNewCommitment(comSig.CommitSig, comSig.HtlcSigs)
+	if err != nil {
+		t.Fatalf("bob failed receiving commitment: %v", err)
+	}
+
+	// Bob sends a new commitment signature, covering htlc2 that he sent
+	// earlier.
+	sig, htlcSigs, err = bobChannel.SignNextCommitment()
+	if err != nil {
+		t.Fatalf("err %v", err)
+	}
+
+	commitSig = &lnwire.CommitSig{
+		CommitSig: sig,
+		HtlcSigs:  htlcSigs,
+	}
+
+	// This signature should cover hltc1 + htlc2.
+	if len(commitSig.HtlcSigs) != 2 {
+		t.Fatalf("expected 2 htlc sigs, got %d",
+			len(commitSig.HtlcSigs))
+	}
+
+	// Alice receives signature covering htlc1 + htlc2 from Bob.
+	aliceLink.HandleChannelUpdate(commitSig)
+
+	// Alice should revoke the old commitment.
+	select {
+	case msg = <-aliceMsgs:
+	case <-time.After(15 * time.Second):
+		t.Fatalf("did not receive message")
+	}
+
+	rev, ok = msg.(*lnwire.RevokeAndAck)
+	if !ok {
+		t.Fatalf("expected RevokeAndAck, got %T", msg)
+	}
+
+	// This will also make Alice ack htlc2.
+	_, _, _, err = bobChannel.ReceiveRevocation(rev)
+	if err != nil {
+		t.Fatalf("bob failed receiving revocation: %v", err)
+	}
+
+	// Bob can now revoke his old commitment in response to the signature
```

Comment:
Where did Alice even send a signature here? If Bob does this, he doesn't even have a valid commitment to play...